### PR TITLE
Add CUP Vehicles Compatibility

### PR DESCRIPTION
**Pull request by Mike-MF**
**When merged this pull request will:**
- Merges CUP Vehicles Compat from CUP to ACE.

ToDo:
- [x] Test it

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    };
+
+    class CUP_MH60S_Base: Helicopter_Base_H {
+        EGVAR(fastroping,enabled) = 1;
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginRight"};
+
+        class UserActions {
+            class OpenDoors;
+            class CloseDoors: OpenDoors {
+                condition = QUOTE(this animationPhase 'doors' > 0.5 && driver this == player && alive (this) && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)]));
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:78]: Functionalize all those conditions please.

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/config.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+class CfgPatches {
+    class SUBADDON {
+        name = COMPONENT_NAME;
+        units[] = {};
+        weapons[] = {};
+        requiredVersion = REQUIRED_VERSION;
+        requiredAddons[] = {
+            "CUP_Vehicles_LoadOrder",
+            "ace_fastroping"
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/config.cpp:11]: I don't see why this is a requirement.

File: addons/compat_cup_vehicles/CfgVehicles.hpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
+class CfgVehicles {
+    class CUP_MTVR_Base;
+    class CUP_MTVR_Reammo_Base: CUP_MTVR_Base {
+        transportAmmo = 0;
```
Comment (PabstMirror) [inline addons/compat_cup_vehicles/CfgVehicles.hpp:4]: I think we can remove rearm/repair of this as it's done dynamically now

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/config.cpp (hunk 2 above)
Comment (PabstMirror) [reply addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/config.cpp:11]: it's for the subaddon

File: addons/compat_cup_vehicles/CfgVehicles.hpp (hunk 3 above)
Comment (jonpas) [reply addons/compat_cup_vehicles/CfgVehicles.hpp:4]: `GVAR(defaultSupply)` should stay unless it inherits the same values.

`transportX` and `GVAR(canRepair)` (if CUP has it configured properly) can be removed.

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    };
+
+    class CUP_UH60_Base: Helicopter_Base_H {
+        EGVAR(fastroping,enabled) = 1;
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginRight"};
+
+        class UserActions {
+            class OpenDoors;
+            class CloseDoors: OpenDoors {
+                condition = QUOTE(this animationPhase 'doors' > 0.5 && driver this == player && alive (this) && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)]));
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:139]:
```suggestion
                condition = QUOTE(this animationPhase 'doors' > 0.5 && driver this == player && alive (this) && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)),false)]));
```

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    class CUP_SA330_Base: Helicopter_Base_H {
+        EGVAR(fastroping,enabled) = 1;
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginLeft", "ropeOriginRight"};
+
+        class UserActions {
+            class CloseRdoor {
+                condition = QUOTE(alive this && {this doorPhase 'ofrp_puma_porte_droite' > 0.5} && {(this getCargoIndex player) isEqualTo 0} && {!(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)])});
+            };
+            class CloseLdoor {
+                condition = QUOTE(alive this && {this doorPhase 'ofrp_puma_porte_gauche' > 0.5} && {(this getCargoIndex player) isEqualTo 1} && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)]));
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:106]:
```suggestion
                condition = QUOTE(alive this && {this doorPhase 'ofrp_puma_porte_gauche' > 0.5} && {(this getCargoIndex player) isEqualTo 1} && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)),false)]));
```

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginLeft", "ropeOriginRight"};
+    };
+
+    class CUP_SA330_Base: Helicopter_Base_H {
+        EGVAR(fastroping,enabled) = 1;
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginLeft", "ropeOriginRight"};
+
+        class UserActions {
+            class CloseRdoor {
+                condition = QUOTE(alive this && {this doorPhase 'ofrp_puma_porte_droite' > 0.5} && {(this getCargoIndex player) isEqualTo 0} && {!(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)])});
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:103]:
```suggestion
                condition = QUOTE(alive this && {this doorPhase 'ofrp_puma_porte_droite' > 0.5} && {(this getCargoIndex player) isEqualTo 0} && {!(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)),false)])});
```

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp (hunk 1 above)
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:78]:
```suggestion
                condition = QUOTE(this animationPhase 'doors' > 0.5 && driver this == player && alive (this) && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)),false)]));
```

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    class CUP_Merlin_HC3_Base: Helicopter_Base_H {
+        EGVAR(fastroping,enabled) = 1;
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginRight"};
+        EGVAR(fastroping,onPrepare) = QEFUNC(compat_cup_vehicles,onPrepareHC3);
+        EGVAR(fastroping,onCut) = QEFUNC(compat_cup_vehicles,onCutHC3);
+
+        class UserActions {
+            class CloseRdoor {
+                condition = QUOTE(this doorPhase 'dvere_p' > 0.5 && {(this getCargoIndex player) isEqualTo 0} && {!(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)])});
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:55]:
```suggestion
                condition = QUOTE(this doorPhase 'dvere_p' > 0.5 && {(this getCargoIndex player) isEqualTo 0} && {!(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)),false)])});
```

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginLeft", "ropeOriginRight", "ropeOriginMid"};
+    };
+
+    class CUP_CH53E_base: Helicopter_Base_H {
+        EGVAR(fastroping,enabled) = 1;
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginLeft", "ropeOriginRight"};
+
+        class UserActions {
+            class RampClose {
+                condition = QUOTE(this animationPhase 'ramp_bottom' >= 0.56 && player == driver this && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)]));
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:42]:
```suggestion
                condition = QUOTE(this animationPhase 'ramp_bottom' >= 0.56 && player == driver this && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)),false)]));
```

File: addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        EGVAR(fastroping,friesType) = "ACE_friesGantry";
+        EGVAR(fastroping,friesAttachmentPoint)[] = {1.05, 1, 1.3};
+        EGVAR(fastroping,onCut) = QEFUNC(fastroping,onCutCommon);
+        EGVAR(fastroping,onPrepare) = QEFUNC(fastroping,onPrepareCommon);
+        EGVAR(fastroping,ropeOrigins)[] = {"ropeOriginLeft", "ropeOriginRight"};
+        EQUIP_FRIES_ATTRIBUTE;
+
+        class UserActions {
+            class CloseRdoor {
+                condition = QUOTE(this doorPhase 'CargoDoorR' > 0.5 && ((this getCargoIndex player) isEqualTo 0) && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)), false)]));
```
Comment (jonpas) [inline addons/compat_cup_vehicles/compat_cup_vehicles_fastroping/CfgVehicles.hpp:26]:
```suggestion
                condition = QUOTE(this doorPhase 'CargoDoorR' > 0.5 && ((this getCargoIndex player) isEqualTo 0) && !(this getVariable [ARR_2(QUOTE(QEGVAR(fastroping,doorsLocked)),false)]));
```

**Review by jonpas — CHANGES_REQUESTED** [review]
Fixing HEMTT warnings.

**Review by PabstMirror — APPROVED** [review]

**Review by LinkIsGrim — APPROVED** [review]

**Review by jonpas — APPROVED** [review]

**Pull request merged**